File: 1605-minimum-number-of-days-to-make-m-bouquets/1605-minimum-number-of-days-to-make-m-bouquets.java
class Solution {
    private boolean f(int[] bloomDay, int day, int m, int k){
        int cnt = 0;
        for(int i : bloomDay){
            if(i <= day){
                cnt++;
                if(cnt == k){
                    m--;
                    cnt = 0;
                    if(m == 0){
                        return true;
                    }
                }
            } else{
                cnt = 0;
            }
        }
        return false;
    }
    public int minDays(int[] bloomDay, int m, int k) {
        int ans = -1, low = 1;
        int high = Arrays.stream(bloomDay).max().getAsInt();
        while(low <= high){
            int mid = low + (high - low) / 2;
            if(f(bloomDay, mid, m, k)){
                ans = mid;
                high = mid - 1;
            } else{
                low = mid + 1;
            }
        }
        return ans;
    }
}

File: 1605-minimum-number-of-days-to-make-m-bouquets/1605-minimum-number-of-days-to-make-m-bouquets.cpp
class Solution {
private:
    bool f(vector<int>& bloomDay, int m, int k, int mid){
        int cnt = 0;
        for(int i : bloomDay){
            if(i > mid){
                cnt = 0;
            } else{
                cnt++;
                if(cnt == k){
                    m--;    cnt = 0;
                }
            }
        }
        return m <= 0;
    }

public:
    int minDays(vector<int>& bloomDay, int m, int k) {
        int l = INT_MAX, r = INT_MIN, ans = -1;
        for(int i : bloomDay){
            l = min(i, l);
            r = max(i, r);
        }

        while(l <= r){
            int mid = l + (r - l) / 2;
            if(f(bloomDay, m, k, mid)){
                ans = mid;
                r = mid - 1;
            } else{
                l = mid + 1;
            }
        }

        return ans;
    }
};